src/main/java/net/sf/mzmine/modules/datapointprocessing/DataPointProcessingController.java

```diff
@@ -0,0 +1,367 @@
+      logger.warning("execute called, without pmc being set.");
+      return;
+    }
+
+    if (getForcedStatus() == ForcedControllerStatus.CANCEL) {
+      setResults(ProcessedDataPoint.convert(dp));
+      logger
+          .finest("Canceled controller, not starting new tasks. Results are set to latest array.");
+      setStatus(ControllerStatus.CANCELED);
+      return;
+    }
+
+    if (step.getModule() instanceof DataPointProcessingModule) {
+
+      DataPointProcessingModule inst = step.getModule();
+      ParameterSet parameters = step.getParameterSet();
+
+      Task t = ((DataPointProcessingModule) inst).createTask(dp, parameters, plot, this,
+          new TaskStatusListener() {
+            @Override
+            public void taskStatusChanged(Task task, TaskStatus newStatus, TaskStatus oldStatus) {
+              if (!(task instanceof DataPointProcessingTask)) {
+                // TODO: Throw exception?
+                logger.warning("This should have been a DataPointProcessingTask.");
+                return;
+              }
+              logger.info("Task status changed to " + newStatus.toString());
+              switch (newStatus) {
+                case FINISHED:
+                  if (pmc.hasNextStep(step)) {
+                    if (DataPointProcessingManager.getInst()
+                        .isRunning(((DataPointProcessingTask) task).getController())) {
+
+                      MZmineProcessingStep<DataPointProcessingModule> next = pmc.getNextStep(step);
+
+                      // pass results to next task and start recursively
+                      ProcessedDataPoint[] result = ((DataPointProcessingTask) task).getResults();
+                      execute(result, next, plot);
+                    } else {
+                      logger.warning(
+                          "This controller was already removed from the running list, although it had not finished processing. Exiting");
+                      break;
+                    }
+                  } else {
+                    setResults(((DataPointProcessingTask) task).getResults());
+                    setStatus(ControllerStatus.FINISHED);
+
+                    displayResults(getResults(), plot);
+
+                  }
+                  break;
+                case PROCESSING:
+                  setStatus(ControllerStatus.PROCESSING);
+                  break;
+                case WAITING:
+                  // should we even set to WAITING here?
+                  break;
+                case ERROR:
+                  setStatus(ControllerStatus.ERROR);
+                  break;
+                case CANCELED:
+                  setStatus(ControllerStatus.CANCELED);
+                  break;
+              }
+            }
+          });
+
+      setCurrentTask((DataPointProcessingTask) t); // maybe we need this some time
+      logger.info("Start processing of " + t.getClass().getName());
+      MZmineCore.getTaskController().addTask(t);

```

Comment:
I think it is using the task controller.
The idea was to prevent any spamming of hundreds of tasks to the task controller

Reply:
Yes, the Manager just "manages" the controllers. It adds them to the task controller when there is no more than (currently) 3 data point tasks running at the moment.